Mathlib/CategoryTheory/Abelian/GrothendieckAxioms.lean

```diff
@@ -0,0 +1,259 @@
+
+/--
+The cocone `finsetBiproductDiagram X` is indeed a colimit.
+This is used to show that `∐ X` is isomorphic to the colimit of `finsetBiproductDiagram X`.
+-/
+@[simps]
+noncomputable
+def isColimitFinsetBiproductCocone : IsColimit (finsetBiproductCocone X) where
+  desc S := Sigma.desc
+    fun a => biproduct.ι (fun (b : ({a} : Finset α)) => X b) ⟨a, by simp⟩ ≫ S.ι.app {a}
+  fac S A := by
+    dsimp
+    ext ⟨a,ha⟩
+    let e : ({a} : Finset α) ⟶ A := homOfLE <| by simpa
+    simp [← S.w e]
+  uniq S m hm := by
+    dsimp
+    ext a
+    simp [← hm {a}]
+
+/--
+The isomorphism between `∐ X` and `finsetBiproductDiagram X`.
+-/
+noncomputable
+def coproductIsoColimitFinsetBiproductDiagram [HasColimitsOfShape (Finset α) C] :
+    ∐ X ≅ colimit (finsetBiproductDiagram X) :=
+  isColimitFinsetBiproductCocone X |>.coconePointUniqueUpToIso <| colimit.isColimit _
+
+@[reassoc (attr := simp)]
+lemma ι_coproductIsoColimitFinsetBiproductDiagram_hom [HasColimitsOfShape (Finset α) C] (a) :
+    Sigma.ι X a ≫ (coproductIsoColimitFinsetBiproductDiagram X).hom =
+    biproduct.ι (fun (b : ({a} : Finset α)) => X b) ⟨a, by simp⟩ ≫
+    colimit.ι (finsetBiproductDiagram X) {a} := by
+  dsimp [coproductIsoColimitFinsetBiproductDiagram, IsColimit.coconePointUniqueUpToIso]
+  simp
+
+/--
+IMPLEMENTATION: Functoriality of `finsetBiproductDiagram`.
+-/
+@[simps]
+noncomputable
+def finsetBiproductNatTrans (η : X ⟶ Y) :
+    finsetBiproductDiagram X ⟶ finsetBiproductDiagram Y where
+  app A := biproduct.map (η ·)
+  naturality {X Y} f := by
+    dsimp
+    ext i j
+    simp only [Category.assoc, biproduct.map_π, biproduct.ι_desc_assoc,
+      ne_eq, biproduct.ι_π_assoc, biproduct.map_desc, biproduct.ι_π]
+    split_ifs with h
+    · subst h
+      simp
+    · simp
+
+variable (α C)
+
+/--
+`finsetBiproductDiagram` as a functor.
+-/
+@[simps]
+noncomputable
+def discreteDiagramToFinsetBiproductDiagram : (Discrete α ⥤ C) ⥤ (Finset α ⥤ C) where
+  obj F := finsetBiproductDiagram fun a => F.obj ⟨a⟩
+  map η := finsetBiproductNatTrans _ _ (η.app ⟨·⟩)
+
+/--
+The coproduct functor indexed by `α` is isomorphic to the composition of
+`discreteDiagramToFinsetBiproductDiagram` and the colimit functor.
+-/
+@[simps!]
+noncomputable
+def coproductFunctorIso [HasCoproducts C] [HasColimitsOfShape (Finset α) C] :
+    (colim (J := Discrete α) (C := C)) ≅ discreteDiagramToFinsetBiproductDiagram C α ⋙ colim :=
+  NatIso.ofComponents
+    (fun F =>
+      HasColimit.isoOfNatIso (Discrete.natIsoFunctor (F := F)) ≪≫
+      coproductIsoColimitFinsetBiproductDiagram (F.obj ⟨·⟩)) <| by
+        intro x y f
+        dsimp
+        ext ⟨⟩
+        simp [Function.comp]
+
+/--
+IMPLEMENTATION: A single component of `discreteDiagramToFinsetBiproductDiagram`.
+-/
+@[simps]
+noncomputable
+def discreteDiagramToBiproduct (A : Finset α) :
+    (Discrete α ⥤ C) ⥤ C where
+  obj F := ⨁ fun a : A => F.obj ⟨a⟩
+  map η := biproduct.map fun _ => η.app _
+
```

Comment:
Up until this point, everything could be done for coproducts instead of biproducts. I imagine that it would be useful to write an arbitrary coproduct as a filtered colimit its finite "sub-coproducts" (and the dual for products). This would need to be done for general cofans rather than the arbitrary chosen coproduct `∐`. Could you try this alternate approach and see if it works?

Reply:
Sounds like a good idea, I'll give it a try.